### PR TITLE
Add test coverage comments to PRs

### PR DESCRIPTION
**Pull request by skpratt**
### Description
In the wake of removing codecov, Consul PRs no longer provide reviewers and contributors with a common reference for test coverage changes. This adds CI commentary on PRs to provide rudimentary coverage metrics and a heatmap to aid reviewers and contributors in evaluating test coverage.

This PR uses a [coverage report generated about a week ago](https://github.com/hashicorp/consul/blob/2cb4988bfecaa32ac9935dca9b22cd8dc3614966/.github/cover.out) to compare coverage changes - therefore this sample coverage report reflects several PRs merged over the past week. Once this code is merged the TEST_COVERAGE_GOLDEN_BRANCH env variable can be updated to point to main, where up-to-date coverage reports will be checked in.

### CODE REVIEW

**Comment by hc-github-team-consul-core** [comment]
All unit tests passed! Your changes bring test coverage over affected code to 63.8%. Total coverage in main is 63.7%. [See coverage delta for your changes](https://output.circle-artifacts.com/output/job/2598cc30-2217-456c-a786-4e4152b77249/artifacts/0/tmp/output/diff_coverage_heatmap.html)

**Review by DanStough — COMMENTED** [review]
🥳 Awesome feature, @skpratt! Some minor questions/comments.

IMHO, I don't love committing back to `main` for TEST_COVERAGE_GOLDEN_BRANCH. It seems to contradict our squash-merging commits to clean up the history. Can we use a dedicated branch, similar to what RB suggested in Slack?

**Comment by DanStough** [comment]
Thanks for the clarification, @skpratt! Does that mean we need to add a filter for this workflow so that it doesn't run on `main` (since the new report will be merged in from the PR)? 

Also, should we disable this for pull requests that run against release branches, since the base code coverage will be wildly different? Or is the plan to have release-branch specific workflows?

**Comment by hc-github-team-consul-core** [comment]
All unit tests passed! Your changes bring test coverage over affected code to 14.7%. Total coverage in main is 63.7%. [See coverage delta for your changes](https://output.circle-artifacts.com/output/job/104c18cd-4691-44c4-b104-9dd78d7a535b/artifacts/0/tmp/output/diff_coverage_heatmap.html)

**Comment by hc-github-team-consul-core** [comment]
All unit tests passed! Your changes bring test coverage over affected code to 14.7%. Total coverage in main is 63.7%. [See coverage delta for your changes](https://output.circle-artifacts.com/output/job/6f0d06be-40d9-4a05-8a41-1d963b242821/artifacts/0/tmp/output/diff_coverage_heatmap.html)

**Comment by skpratt** [comment]
> Thanks for the clarification, @skpratt! Does that mean we need to add a filter for this workflow so that it doesn't run on `main` (since the new report will be merged in from the PR)?
> 
> Also, should we disable this for pull requests that run against release branches, since the base code coverage will be wildly different? Or is the plan to have release-branch specific workflows?

Great call out, @DanStough! I just exempted these branches, because overwriting coverage on main would create noise, and coverage comparison would indeed be meaningless on release branches (we can maybe consider a similar approach for release branches if this experiment works out).

**Comment by hc-github-team-consul-core** [comment]
All unit tests passed! Your changes bring test coverage over affected code to 43.9%. Total coverage in main is 63.7%. [See coverage delta for your changes](https://output.circle-artifacts.com/output/job/e4a90641-8d10-468c-82cb-531f275e8718/artifacts/0/tmp/output/diff_coverage_heatmap.html)

**Comment by hc-github-team-consul-core** [comment]
All unit tests passed! Your changes bring test coverage over affected code to 49.0%. Total coverage in main is 63.7%. [See coverage delta for your changes](https://output.circle-artifacts.com/output/job/21ad9e08-e153-4ed7-9be4-3c6d22a5b78b/artifacts/0/tmp/output/diff_coverage_heatmap.html)

File: .circleci/config.yml [hunk 1]
```diff
@@ -1036,6 +1024,81 @@ jobs:
...
+            mkdir -p /tmp/output
+            echo "mode: set" > cov.txt
+            for f in /tmp/coverage/*; do tail -n +2 $f >> cov.txt; done
+            go tool cover -func=cov.txt -o /tmp/output/coverage_by_function.txt
+            cp cov.txt /tmp/output/cov.txt
+      - run: sudo apt-get install jq
+      - run:
+          name: get diff of test coverage from changes
+          command: |
+            file_url_response=$(curl --location --request GET "https://api.github.com/repos/hashicorp/consul/contents/.github/cover.out?ref=$TEST_COVERAGE_GOLDEN_BRANCH" \
```
Comment (dhiaayachi) [inline .circleci/config.yml:1046]: I think it's better to extract the URL into a variable to make it easier to port to enterprise.

File: .circleci/config.yml [hunk 2]
```diff
@@ -1036,6 +1024,81 @@ jobs:
...
+            echo "mode: set" > diff_cov.txt
+            comm -13 <(sort main_cov.txt) <(sort cov.txt) >> diff_cov.txt
+            go tool cover -html=diff_cov.txt -o /tmp/output/diff_coverage_heatmap.html
+      - store_artifacts:
+          path: /tmp/output
+      - run:
+          name: post results to GitHub PR
+          command: |
+            pr_response=$(curl --location --request GET "https://api.github.com/repos/hashicorp/consul/pulls?head=hashicorp:$CIRCLE_BRANCH&state=open" \
+            -u $GH_USER:$GH_TOKEN)
```
Comment (dhiaayachi) [inline .circleci/config.yml:1065]: Are those token added as secrets?

File: .circleci/config.yml [hunk 3]
```diff
@@ -1036,6 +1024,81 @@ jobs:
...
+            else
+              pr_comment_url=$(echo $pr_response | jq -r ".[]._links.comments.href")
+            fi
+            
+            new_cov_percent=$(tail -1 /tmp/output/coverage_by_function.txt | awk -F' ' '{print $NF}')
+            go tool cover -func=main_cov.txt -o main_cov_by_func.txt
+            main_go_test_cov_percent=$(tail -1 main_cov_by_func.txt | awk -F' ' '{print $NF}')
+            artifacts=$(curl -X GET "https://circleci.com/api/v2/project/github/hashicorp/consul/$CIRCLE_BUILD_NUM/artifacts" \
+            -H "Accept: application/json" \
+            -u "$CIRCLE_API_TOKEN:")
```
Comment (dhiaayachi) [inline .circleci/config.yml:1078]: Is this added as a secret?

**Review by dhiaayachi — APPROVED** [review]
Thank you for working on this @skpratt. I added few comments and questions but nothing blocking.

I still think this would be easier if done using github actions and an S3 bucket or artifactory. It will remove all the complexity related to git and GH/Circle API access and the drawback of having to carry an extra coverage file with each PR.
That said, we can maybe merge as is and pair later on a newer version.

**Review by DanStough — APPROVED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (skpratt) [reply .circleci/config.yml]: extracted project into variable for ease of porting

File: .circleci/config.yml (hunk 3 above)
Comment (skpratt) [reply .circleci/config.yml:1079]: yes

**Comment by skpratt** [comment]
> Thank you for working on this @skpratt. I added few comments and questions but nothing blocking.
> 
> I still think this would be easier if done using github actions and an S3 bucket or artifactory. It will remove all the complexity related to git and GH/Circle API access and the drawback of having to carry an extra coverage file with each PR. That said, we can maybe merge as is and pair later on a newer version.

I fully agree! I see this an initial test to assess the value of these metrics. If we decide to keep it, the next step would be to shift the infra, and I'd love to pair on that with you or anyone else who may be interested.

**Pull request merged**